tcmur_device.c

```diff
@@ -207,26 +207,26 @@ void tcmu_notify_lock_lost(struct tcmu_device *dev)
 int tcmu_cancel_lock_thread(struct tcmu_device *dev)
 {
 	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
-	void *join_retval;
 	int ret;
 
 	pthread_mutex_lock(&rdev->state_lock);
 	if (rdev->lock_state != TCMUR_DEV_LOCK_LOCKING) {
 		pthread_mutex_unlock(&rdev->state_lock);
 		return 0;
 	}
-	pthread_mutex_unlock(&rdev->state_lock);
 	/*
 	 * It looks like lock calls are not cancelable, so
 	 * we wait here to avoid crashes.
 	 */
 	tcmu_dev_dbg(dev, "Waiting on lock thread\n");
-	ret = pthread_join(rdev->lock_thread, &join_retval);
+
 	/*
 	 * We may be called from the lock thread to reopen the device.
 	 */
-	if (ret != EDEADLK)
-		tcmu_dev_err(dev, "pthread_join failed with value %d\n", ret);
+	tcmu_dev_dbg(rdev->dev, "waiting for lock thread to exit\n");
+	ret = pthread_cond_wait(&rdev->lock_cond, &rdev->state_lock);
+	pthread_mutex_unlock(&rdev->state_lock);
+
```

Comment:
I think we can hit a deadlock here. See the comment above about if we are called from tcmu_acquire_dev_lock -> tcmu_reopen_dev -> tcmu_cancel_lock_thread. In this case we would get stuck in the wait call above.  I can send a patch over this though. Will update this with the PR.

Reply:
Yes, just see your this reply.
I have pushed a new one. but still has some problems.

Reply:
>  I have pushed a new one. but still has some problems.

What issues are you still hitting? Is it a bug added in this patch or did you mean something related to the issue https://github.com/open-iscsi/tcmu-runner/pull/332
?

I tested the patch real quick here and it seems ok.

Reply:
Sorry, I think I have just copied the wrong rpm packages to do the test.
And it works fine on my RHEL 7.2(this is very easy to reproduce this bug) and Centos environments.